subsys/net/l2_wifi_if_conn/l2_wifi_conn.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(l2_wifi_mgr_conn, CONFIG_NET_CONNECTION_MANAGER_LOG_LEVEL);
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/wifi_mgmt.h>
+#include <zephyr/net/net_event.h>
+#include <net/wifi_mgmt_ext.h>
+#include <net/l2_wifi_connect.h>
+#include <zephyr/net/conn_mgr_connectivity.h>
+
+#define WIFI_SHELL_MGMT_EVENTS (NET_EVENT_WIFI_CONNECT_RESULT | \
+				NET_EVENT_WIFI_DISCONNECT_RESULT)
+
+static struct net_mgmt_event_callback net_l2_mgmt_cb;
+int connection_status;
+
+K_SEM_DEFINE(wait_for_conn_event_sem, 0, 1);
+
+static void net_l2_wifi_connect_result(struct net_mgmt_event_callback *cb)
+{
+	k_sem_give(&wait_for_conn_event_sem);
+}
+
+static void net_l2_wifi_mgmt_event_handler(struct net_mgmt_event_callback *cb,
+				    uint32_t mgmt_event, struct net_if *iface)
+{
+	switch (mgmt_event) {
+	case NET_EVENT_WIFI_CONNECT_RESULT:
+		net_l2_wifi_connect_result(cb);
+		break;
+	case NET_EVENT_WIFI_DISCONNECT_RESULT:
+		break;
+	default:
+		break;
+	}
+}
+
+int net_l2_wifi_connect(struct conn_mgr_conn_binding *const binding)
+{
+	uint64_t time_stamp_start;
+	uint32_t ticks_elapsed;
+	int rc;
+	int sem_ret;
+
+	time_stamp_start = (uint64_t)k_uptime_ticks();
+	if (!binding) {
+		return -EINVAL;
+	}
+
+	while (true) {
+		rc = net_mgmt(NET_REQUEST_WIFI_CONNECT_STORED, binding->iface, NULL, 0);
+
+		if (binding->flags && CONN_MGR_IF_PERSISTENT != CONN_MGR_IF_PERSISTENT) {
+			break;
+		}
+
+		/* wait here for connection status */
+		sem_ret = k_sem_take(&wait_for_conn_event_sem, K_MSEC(5000));
+		if (sem_ret) {
+			NET_ERR("%s: Failed to get the status, ret = %d\n", __func__, sem_ret);
+			rc = -1;
+			break;
+		}
+
+		/* Break if connection was successful */
+		if (!((const struct wifi_status *) &net_l2_mgmt_cb.info)->status) {
+			break;
+		}
+
+		ticks_elapsed = k_ticks_to_ms_ceil32((uint64_t)k_uptime_ticks() - time_stamp_start);
+		if (binding->timeout && (ticks_elapsed >= ((binding->timeout)*1000))) {
+			break;
+		}
```

Comment:
I recommend using a workqueue for handing timeouts, I think making `net_l2_wifi_connect` non-blocking will make them the most convenient solution. Here is a good example from the LTE integration:

https://github.com/nrfconnect/sdk-nrf/pull/10177/commits/1c9c64fc20c7b563b898e49009a6a5cb2d2eb112

Specifically check out `lib/nrf_modem_lib/lte_connectivity/lte_connectivity.c` and `connection_timeout_schedule`, `connection_timeout_work_fn`, so on

In your case, things are a bit more complicated since you need to have concurrent retry logic.

You can probably use a single workqueue with two workqueue items: Timeout and Retry, where the "timeout" work item cancels all ongoing connection attempts, and "Retry" would be a work item that performs one loop of this function, but instead of `k_sem_take`, it just schedules itself to reoccur in 5 seconds.

Reply:
Implemented work_queue to make connect_call non-blocking. Looking into a possible race condition between reading and updating of connection status.